src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will let one to hold onto <see cref="RemoteHostClient.Session"/> for a while.
+    /// this helper will let you not care about remote host being gone while you hold onto the session if that ever happen
+    /// 
+    /// and also make sure state is correct even if multiple threads call TryInvokeAsync at the same time. but this 
+    /// is not optimized to handle highly concurrent usage. if highly concurrent usage is required, either using
+    /// <see cref="RemoteHostClient.Session"/> direclty or using <see cref="SolutionAndSessionHolder"/> would be better choice
+    /// </summary>
+    internal sealed class KeepAliveSessionHolder
+    {
+        private readonly SemaphoreSlim _gate;
+        private readonly IRemoteHostClientService _remoteHostClientService;
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly string _serviceName;
+        private readonly object _callbackTarget;
+
+        private RemoteHostClient _client;
+        private RemoteHostClient.Session _session;
+
+        public KeepAliveSessionHolder(RemoteHostClient client, RemoteHostClient.Session session, string serviceName, object callbackTarget, CancellationToken cancellationToken)
+        {
+            Initialize_NoLock(client, session);
+
+            _gate = new SemaphoreSlim(initialCount: 1);
+            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
+            _cancellationToken = cancellationToken;
+
+            _serviceName = serviceName;
+            _callbackTarget = callbackTarget;
+        }
+
+        public void Shutdown()
```

Comment:
> Their behavior isn't different. 
They are different. 

> They get the session. The call operations on it. They dispose the session once they're done with it.
Ya, They get a type. They call operations on it. They dispose the type once they're done with it. why does it have to be named session?

> And that's evident in that you have to expose the same methods on all your new types.

they are not same shape. one tie session to one solution. the other doesn't tie session to one solution. one exist people who doesn't want to put solution to each of their Invoke. the other exist for people who want to keep using same session for multiple solutions.

also, one is designed to be use with "using" and the other is supposed to not used with "using".

also KeepAlive one implements handling for OOP restart case which we need for nuget analyzer being reloaded without restart VS.

they are 2 different thing. only thing they resembles are they are for OOP and they have same methods name InvokeAsync which actually have different signature.